File: CTCI/JavaScript/01_Arrays_and_Strings/07_rotate_matrix.js
/**
 * Problem:
 * Given an image represented by an NxN matrix, where each pixel in the image
 * is 4 bytes, write a method to rotate the image by 90 degrees. Can you do this
 * in place?
 * 
 * Example:
 *    Input:    |   Output:
 * [            | [
 *  [1, 2, 3],  |   [7, 4, 1],
 *  [4, 5, 6],  |   [8, 5, 2],
 *  [7, 8, 9]   |   [9, 6, 3]
 * ]            | ]
 * 
 * @param  {array} matrix Matrix to rotate
 * @return {array}        Rotated input matrix
 */

function rotateMatrix(matrix) {
  if (!matrix || matrix.length === 0 || matrix.length !== matrix[0].length) {
    throw new Error('invalid matrix');
  }
  // a rotated 1x1 matrix is the same
  if (matrix.length === 1) return matrix;
  for (let i = 0; i < matrix.length; i++) {
    for (let j = i + 1; j < matrix[i].length; j++) {
      let temp = matrix[i][j];
      matrix[i][j] = matrix[j][i];
      matrix[j][i] = temp;
    }
  }
  for (let i = 0; i < matrix.length; i++) {
    matrix[i].reverse();
  }
}

File: CTCI/Java/01_Arrays_and_Strings/07_rotate_matrix.java
/**
 * Problem:
 * Given an image represented by an NxN matrix, where each pixel in the image
 * is 4 bytes, write a method to rotate the image by 90 degrees. Can you do this
 * in place?
 * 
 * Example:
 *    Input:    |   Output:
 * [            | [
 *  [1, 2, 3],  |   [7, 4, 1],
 *  [4, 5, 6],  |   [8, 5, 2],
 *  [7, 8, 9]   |   [9, 6, 3]
 * ]            | ]
 * 
 * Time Complexity: O(N^2), must access all N^2 elements of the matrix
 *
 * @param matrix Matrix to rotate
 * @return Rotated input matrix
 */

class RotateMatrix {
    boolean rotateMatrix(int[][] matrix) {
        // check if matrix is empty or if matrix has NxN dimensions
        if (matrix.length == 0 || matrix.length != matrix[0].length) {
            return false;
        }

        int n = matrix.length;
        /* Go layer by layer, starting w/ the outer layer and working inwards.
         * Perform a swap on each layer */
        for (int layer = 0; layer < n / 2; layer++) {
            int first = layer;
            int last = n - 1 - layer;
            for (int i = first; i < last; i++) {
                int offset = i - first;
                // save the top
                int top = matrix[first][i];

                // left -> top
                matrix[first][i] = matrix[last - offset][first];

                // bottom -> left
                matrix[last - offset][first] = matrix[last][last - offset];

                // right -> bottom
                matrix[last][last - offset] = matrix[i][last];

                // top -> right (saved top)
                matrix[i][last] = top;
            }
        }
        return true;
    }
}